pkg/sql/revert_test.go

```diff
@@ -106,6 +111,206 @@ func TestTableRollback(t *testing.T) {
+
+	// Delete some rows
+	db.Exec(t, `DELETE FROM test WHERE k % 5 = 2`)
+	db.Exec(t, `DELETE FROM test2 WHERE k % 5 = 2`)
+
+	afterDelete1, _ := fingerprintTableNoHistory(t, db, tableID1, "")
+	afterDelete2, _ := fingerprintTableNoHistory(t, db, tableID2, "")
+	require.NotEqual(t, before1, afterDelete1)
+	require.NotEqual(t, before2, afterDelete2)
+
+	aost1, _ := fingerprintTableNoHistory(t, db, tableID1, ts)
+	aost2, _ := fingerprintTableNoHistory(t, db, tableID2, ts)
+	require.Equal(t, before1, aost1)
+	require.Equal(t, before2, aost2)
+
+	const ignoreGC = false
+
+	// Bypass the offline check.
+	desc1.TableDesc().State = descpb.DescriptorState_OFFLINE
+	desc2.TableDesc().State = descpb.DescriptorState_OFFLINE
+
+	require.NoError(t, sql.RevertTables(context.Background(), kv, &execCfg, []catalog.TableDescriptor{desc1, desc2}, targetTime, ignoreGC, 10))
+
+	reverted1, _ := fingerprintTableNoHistory(t, db, tableID1, "")
+	reverted2, _ := fingerprintTableNoHistory(t, db, tableID2, "")
+	require.Equal(t, before1, reverted1, "expected reverted table after edits to match before")
+	require.Equal(t, before2, reverted2, "expected reverted table after edits to match before")
+}
+
+// TestRevertSpansFanout tests RevertSpansFanout using the same
+// sequence of modifications used in TestTableRollback.
+//
+// We use more splits and scatter the ranges.
+func TestRevertSpansFanout(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+
+	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{UseDatabase: "test"},
+	})
+	defer tc.Stopper().Stop(context.Background())
+	s := tc.TenantOrServer(0)
+	sqlDB := tc.Conns[0]
+
+	execCfg := s.ExecutorConfig().(sql.ExecutorConfig)
+
+	db := sqlutils.MakeSQLRunner(sqlDB)
+	db.Exec(t, "CREATE DATABASE IF NOT EXISTS test")
+	db.Exec(t, "CREATE TABLE test (k INT PRIMARY KEY, rev INT DEFAULT 0, INDEX (rev))")
+
+	kvDB := s.DB()
+	desc := desctestutils.TestingGetPublicTableDescriptor(kvDB, execCfg.Codec, "test", "test")
+	span := desc.TableSpan(execCfg.Codec)
+	tableID := desc.GetID()
+
+	// Fill a table with some rows plus some revisions to those rows.
+	const (
+		initRows  = 1000
+		extraRows = 500
+		splits    = 100
+	)
+
+	db.Exec(t, "INSERT INTO test (k) SELECT generate_series(1, $1)", initRows)
+	db.Exec(t, "UPDATE test SET rev = 1 WHERE k % 3 = 0")
+	db.Exec(t, "DELETE FROM test WHERE k % 10 = 0")
+	before, ts := fingerprintTableNoHistory(t, db, tableID, "")
+	targetTime, err := hlc.ParseHLC(ts)
+	require.NoError(t, err)
+
+	beforeNumRows := db.QueryStr(t, "SELECT count(*) FROM test")
+
+	// Make some more edits: delete some rows and edit others, insert into some of
+	// the gaps made between previous rows, edit a large swath of rows and add a
+	// large swath of new rows as well.
+	db.Exec(t, "DELETE FROM test WHERE k % 5 = 2")
+	db.Exec(t, "INSERT INTO test (k, rev) SELECT generate_series(10, $1, 10), 10", initRows)
```

Comment:
was this line suppose to use `extraRows`? It seems like `extraRows` was never used to insert stuff.

Reply:
Thanks, I meant to parameterize the hard coded 500 below.